Add a new concept page for service accounts

shannonxtreme opened the pull request:
Add a new concept page for Service Accounts in k8s

This PR focuses only on adding the new page, based on the draft doc [here](https://docs.google.com/document/d/1UcEFreczITmFhcHAgZkIpuEJD4m1AraiRQAtzpNIRaE/edit#). It covers key topics like use cases, defaults, and authentication. 

Some additional work after this:

1. Clean up existing pages to remove duplicate info if exists.
1. Link from how-tos back to the concept
1. Move information about ServiceAccount automation from https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#control-plane-details to this page

Preview: https://deploy-preview-38289--kubernetes-io-main-staging.netlify.app/docs/concepts/security/service-accounts/

Issue: #31847

/sig docs
/language en

shannonxtreme commented:
/cc @sftim @liggitt @Shubham82 @kubernetes/sig-auth-pr-reviews

sftim reviewed (approved):
/approve

[JWTs do not necessarily include a digital signature](https://datatracker.ietf.org/doc/html/rfc7519#section-6), so please fix that.

/hold
pending a change in light of the JWT feedback I just wrote

/thanks
Much appreciated!

shannonxtreme commented:
@sftim I modified the wording slightly to " A JSON object that can be digitally signed and used for authentication."

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+ServiceAccounts use signed
+{{<glossary_tooltip term_id="jwt" text="JSON Web Tokens (JWTs)">}}
+to authenticate to the Kubernetes API, and to any other system where a trust
```
```suggestion
to authenticate to the Kubernetes API server, and to any other system where a trust
```

shannonxtreme replied:
Done

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+Kubernetes API server, it adds an `Authorization: Bearer <token>` header to the
+HTTP request. The API server checks the validity of the bearer token as follows:
+
+1.   Check the token signature.
+2.   Check whether the token has expired.
+3.   Check whether the token is currently valid.
+4.   Check the audience claims.
```
```suggestion
1. Check the token signature.
2. Check whether the token has expired.
3. Check whether the token is currently valid.
4. Check the audience claims.
```

shannonxtreme replied:
Done (and fixed markdown numbering)

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+mounted as Secrets in Pods, the API server checks the token against the Secret.
+
+For more information about the authentication process, refer to
+[Authenticating](/docs/reference/access-authn-authz/authentication/#service-account-tokens).
```
```suggestion
[Authentication](/docs/reference/access-authn-authz/authentication/#service-account-tokens).
```

shannonxtreme replied:
Done

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+If you have services of your own that need to validate Kubernetes service
+account credentials, you can use the following methods:
+
+*   [TokenReview API](/docs/reference/kubernetes-api/authentication-resources/token-review-v1/)
+    (recommended)
+*   OIDC discovery
```
```suggestion
* [TokenReview API](/docs/reference/kubernetes-api/authentication-resources/token-review-v1/)
  (recommended)
* OIDC discovery
```

shannonxtreme replied:
Done

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+## Alternatives
+
+*   Issue your own tokens using another mechanism, and then use
+    [Webhook Token Authentication](/docs/reference/access-authn-authz/authentication/#webhook-token-authentication)
+    to validate bearer tokens using your own validation service.
+*   Provide your own identities to Pods.
+    *   [Use the SPIFFE CSI driver plugin to provide SPIFFE SVIDs as X.509 certificate pairs to Pods](https://cert-manager.io/docs/projects/csi-driver-spiffe/).
+    *   [Use a service mesh such as Istio to provide certificates to Pods](https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/).
+*   Authenticate from outside the cluster to the API server without using service account tokens:
+    *   [Configure the API server to accept OpenID Connect (OIDC) tokens from your identity provider](/docs/reference/access-authn-authz/authentication/#openid-connect-tokens).
+    *   Use service accounts or user accounts created using an external Identity
+        and Access Management (IAM) service, such as from a cloud provider, to
+        authenticate to your cluster.
```
```suggestion
* Issue your own tokens using another mechanism, and then use
  [Webhook Token Authentication](/docs/reference/access-authn-authz/authentication/#webhook-token-authentication)
  to validate bearer tokens using your own validation service.
* Provide your own identities to Pods.
  * [Use the SPIFFE CSI driver plugin to provide SPIFFE SVIDs as X.509 certificate pairs to Pods](https://cert-manager.io/docs/projects/csi-driver-spiffe/).
  * [Use a service mesh such as Istio to provide certificates to Pods](https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/).
* Authenticate from outside the cluster to the API server without using service account tokens:
  * [Configure the API server to accept OpenID Connect (OIDC) tokens from your identity provider](/docs/reference/access-authn-authz/authentication/#openid-connect-tokens).
  * Use service accounts or user accounts created using an external Identity
    and Access Management (IAM) service, such as from a cloud provider, to
    authenticate to your cluster.
```

shannonxtreme replied:
Done

tengqm commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+## {{% heading "whatsnext" %}}
+
+*   [Learn how to manage your ServiceAccounts as a cluster administrator](/docs/reference/access-authn-authz/service-accounts-admin/).
+*   [Learn how to assign a ServiceAccount to a Pod](/docs/tasks/configure-pod-container/configure-service-account/).
+*   [Read the ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```
```suggestion
* Learn how to [manage your ServiceAccounts as a cluster administrator](/docs/reference/access-authn-authz/service-accounts-admin/).
* Learn how to [assign a ServiceAccount to a Pod](/docs/tasks/configure-pod-container/configure-service-account/).
* Read [the ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```

shannonxtreme replied:
Done

shannonxtreme replied:
Actually done

tengqm commented on content/en/docs/reference/glossary/jwt.md:
```diff
@@ -0,0 +1,19 @@
+
+<!--more-->
+
+Kubernetes uses JSON Web Tokens (JWTs) to verify the identity of objects
```
```suggestion
Kubernetes can use JSON Web Tokens (JWTs) to verify the identity of objects
```

shannonxtreme replied:
Done

sftim commented on content/en/docs/reference/glossary/jwt.md:
```diff
@@ -0,0 +1,19 @@
+date: 2022-12-06
+full_link: https://www.rfc-editor.org/rfc/rfc7519
+short_description: >
+  A digitally signed JSON object that can be used for authentication.
```
```suggestion
  An authentication token format based on JSON data.
```

shannonxtreme replied:
Done

sftim replied:
This suggestion is pending.

shannonxtreme replied:
Weird, thought I committed it. done now.

sftim commented on content/en/docs/reference/glossary/jwt.md:
```diff
@@ -0,0 +1,19 @@
+- security
+- architecture
+---
+ A JSON object that can be digitally signed and used for authentication.
```
BTW, JWTs are _not_ JSON objects. Usually they contain three separate JSON objects in a [custom format](https://www.rfc-editor.org/rfc/rfc7516#page-8).

shannonxtreme replied:
Applied your suggestion. is the "digitally signed" thing needed?

sftim replied:
I'd include it. JWT signing is an important feature.

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+## What are service accounts?
+
+A service account is a type of non-human account that, in Kubernetes, provides
+an identity for the processes that run in Pods. Pods can use service accounts
```
processes in pods are the primary use, but are not the only one... service account credentials can be obtained and used outside of pods as well (notably, the kubernetes controller manager uses a distinct service account for every controller loop)

shannonxtreme replied:
Done

sftim replied:
Not sure how this was resolved yet - sorry. We can suggest wording if that'll help.

shannonxtreme replied:
would something like this work?

```suggestion
an identity for cluster processes, such as Pods and controllers, to use when talking to the API server
```

sftim replied:
> ```markdown
> A service account is a type of non-human account that, in Kubernetes, provides
> an identity for cluster processes, such as Pods and controllers, to use when talking to the API server
> to make calls to the Kubernetes API server. Service accounts exist as
> ```

It's not quite that. You can also use a ServiceAccount for things other than authenticating to the API server,
so we need to frame this authn piece as only one part of the picture.

For example, you can use a ServiceAccount to define which image pull secrets a Pod can use. ServiceAccounts are really there to define an identity (≅ authentication principal) for _sets_ of Pods. Those sets of Pods might be the same as an app component, or:
- one app might use several different ServiceAccounts, for different things
- several apps might all use the same ServiceAccount

sftim replied:
When Pods use a ServiceAccount, the Pods and the ServiceAccount have to be in the same namespace (this isn't implemented as a check, it's just - by design -  the only way the API lets you specify it).

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+*  **Namespaced:** Each service account is bound to a Kubernetes
+   {{<glossary_tooltip text="namespace" term_id="namespace">}}. Every namespace
+   gets a `default` ServiceAccount upon creation.
```
is it clear that `default` is the actual name of the service account?

shannonxtreme replied:
Rephrased

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+   the configurations portable.
+
+Service accounts are different from user accounts, which are authenticated
+human users in the cluster. User accounts don't exist in the Kubernetes API
+server; instead, the API server treats user identities as opaque data. You can
```
```suggestion
human users in the cluster. User accounts are not represented as Kubernetes API
objects; instead, the API server treats user identities as opaque data. You can
```

shannonxtreme replied:
Done

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+When you create a cluster, Kubernetes automatically creates a ServiceAccount
+object named `default` for every namespace in your cluster. The `default`
+service accounts in each namespace get no permissions other than the
```
they can be given other permissions, but get none by default:

```suggestion
service accounts in each namespace get no permissions by default other than the
```

shannonxtreme replied:
Done

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+If you deploy a Pod in a namespace, and you don't manually assign a
+ServiceAccount object to the Pod, Kubernetes assigns the `default` service
+account for that namespace to the Pod. You can change this behavior by setting
+the `automountServiceAccountToken` field in your Pod specification to `false`.
```
this isn't quite accurate... the `default` service account still gets assigned even when `autoMountServiceAccountToken` is false, but no token is auto-mounted.

shannonxtreme replied:
Done

shannonxtreme replied:
Actually done, idk where these changes disappeared to

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+You can use RBAC to allow service accounts in one namespace to perform actions
+on resources in a different namespace in the cluster. For example, consider a
+scenario where you have a service account and Pod in the `dev` namespace and
+you want your Pod to see Jobs running in the `batch` namespace. You could
```
nit: to avoid confusion, maybe use a different example name than the API group for jobs?

```suggestion
you want your Pod to see Jobs running in the `maintenance` namespace. You could
```

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+on resources in a different namespace in the cluster. For example, consider a
+scenario where you have a service account and Pod in the `dev` namespace and
+you want your Pod to see Jobs running in the `batch` namespace. You could
+create a Role object that grants permissions to list Job objects. Then,
```
```suggestion
create a Role object in the `maintenance` namespace that grants permissions to list Job objects. Then,
```

shannonxtreme replied:
Ignore the last comment, the Codespaces review thing took me in reverse order

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+you want your Pod to see Jobs running in the `batch` namespace. You could
+create a Role object that grants permissions to list Job objects. Then,
+you'd create a RoleBinding object in the `batch` namespace to bind the Role
+to the ServiceAccount object. Now, Pods in the `dev` namespace can list Job
```
```suggestion
to the ServiceAccount object in the `dev` namespace. Now, Pods in the `dev` namespace can list Job
```

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+### Use ServiceAccount credentials from a Pod
+
+To assign a service account to a Pod, you provide the Pod with the credentials
+for the service account, which the Pod can then use to authenticate with the
```
from a user-perspective, you assign a service account to a pod by setting `spec.serviceAccountName`; the system does the injection (using one of the following methods) automatically. I probably wouldn't mention any of the specific methods this in this section... it's more relevant in a "Use ServiceAccount credentials outside a Pod" section

shannonxtreme replied:
Ah I see. https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount cleared it up a bit. So depending on k8s version, the method by which Kubernetes gets SA credentials is different? In versions earlier than 1.22, tokens were autogenerated for Secret-based mounting. In 1.22 and later, Kubernetes gets the tokens using TokenRequest and mounts as projected volumes. When would you use those manually then? tokenrequest I understand, to get SA credentials in an application, but what about manually mounting as a projected volume? 

Sorry @liggitt I dunno if my response is very coherent

liggitt replied:
> I'd still need to do something in my Pod spec to tell Kubernetes that I want a projected volume, for example, right?

No, auto-injection of a token for the pod's service account silently switched mechanisms from secret-based to projected-volume-based in ~1.21/1.22.

> Or is it automatic based on version and the only thing I need to set is the ServiceAccount name?

Yeah, pod authors just set the spec.serviceAccountName, API server admission handles the rest of the volume setup.

> When would you use those manually then? tokenrequest I understand, to get SA credentials in an application, but what about manually mounting as a projected volume?

If you want a service account token mounted at a non-standard location, or with an audience other than the kube-apiserver, you would manually set up a projected token mount. That would let you get a token you could present to some other audience (for example, scenarios mentioned in https://github.com/cert-manager/cert-manager/issues/4144#issuecomment-930087333)

shannonxtreme replied:
I'll fiddle with this today, this is the last bit I think is needed to get this PR in good order

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+    is not recommended, especially at scale, because of the risks associated
+    with static, long-lived credentials.
+
+## Authenticating service account credentials
```
nit: I was confused by the title into thinking this was describing how to validate credentials, but that's the next section. Maybe a title tweak like this:

```suggestion
## Authenticating using service account credentials
```

Shubham82 commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+* Your Pods need to communicate with an external service. For example,
+  [Authenticating to a private image registry using an `imagePullSecret`](/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account).
+* An external service needs to communicate with the Kubernetes API server. For
+  example, authenticating to the cluster as part of a CI/CD pipeline.
```
Is it better if we have some references for the external service use cases? we can make a section of it such as `Use ServiceAccount credentials from external`. It will be better to understand the things

shannonxtreme replied:
We can do that, but from what I see right now we only have one defined use case. Maybe after we publish this you'd be willing to open a new issue or PR with more to add?

Shubham82 replied:
I would like to know that use case, could you please refer that link to me.

Shubham82 replied:
Are you talking about **authenticating to the cluster as part of a CI/CD pipeline**?

shannonxtreme replied:
Yeah @Shubham82 that's the only example I have for this PR. We can always expand with more examples down the line though

Shubham82 replied:
> We can always expand with more examples down the line though

Sure
Thanks, @shannonxtreme for this.

Shubham82 commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+    (not recommended): You can mount service account tokens as Kubernetes
+    Secrets in Pods. These tokens don't expire and don't rotate. This method
+    is not recommended, especially at scale, because of the risks associated
+    with static, long-lived credentials.
```
we can mention the note for `manually created secret-based tokens` about the depreciation plan. IMO  there is no current plan to deprecate manually created secret-based tokens. so can we mention this thing also?

shannonxtreme replied:
I'm confused here Shubham, what do you mean by the deprecation plan?

shannonxtreme replied:
@tengqm would a summarized version of the note in https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount be good here?

Shubham82 replied:
> I'm confused here Shubham, what do you mean by the deprecation plan?

I meant that, Is there any plan to deprecate `manually created secret-based tokens`in future?

Shubham82 replied:
> I'm confused here Shubham, what do you mean by the deprecation plan?
> 
> I meant that, Is there any plan to deprecate manually created secret-based tokensin future?

What about this one @shannonxtreme,

shannonxtreme replied:
I do not think that there's a plan to deprecate this functionality as of right now. I'll implement Qimings suggestion. Thanks for reminding me about this one Shubham, when there's so many comments it's easy to miss one 😂

Shubham82 replied:
> I do not think that there's a plan to deprecate this functionality as of right now. I'll implement Qimings suggestion.

Yeah, sounds good.

Shubham82 replied:
> Thanks for reminding me about this one Shubham, when there's so many comments it's easy to miss

Welcome, I understand it.

Shubham82 replied:
> > IMO there is no current plan to deprecate manually created secret-based tokens
> 
> These are still supported. _Automatic_ generation is already deprecated: [kubernetes/enhancements#2799](https://github.com/kubernetes/enhancements/issues/2799)

Thanks @sftim for it, so can we mention the feature gate that stops Automatic generation or service accounts? and also mention **manually created secret-based tokens** that no plan to remove in the future?

Shubham82 replied:
@shannonxtreme WDYT?

mehabhalodiya commented:
@shannonxtreme can you please do the suggested changes? thanks!

shannonxtreme commented:
> @shannonxtreme can you please do the suggested changes? thanks!

I will when I'm back in the groove, currently taking a much needed couple of days to myself Meha ☺️

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+object named `default` for every namespace in your cluster. The `default`
+service accounts in each namespace get no permissions other than the
+[default API discovery permissions](/docs/reference/access-authn-authz/rbac/#service-account-permissions)
+that Kubernetes grants to all authenticated accounts. If you delete the default
```
Done

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+If you deploy a Pod in a namespace, and you don't manually assign a
+ServiceAccount object to the Pod, Kubernetes assigns the `default` service
+account for that namespace to the Pod. You can change this behavior by setting
+the `automountServiceAccountToken` field in your Pod specification to `false`.
```
Updated by saying how you might not specify a service account for the Pod in line 65-66

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+
+To use a Kubernetes service account, you do the following:
+
+1. Create the ServiceAccount object in the Kubernetes API using a Kubernetes
```
Done

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+use that service account don't get more permissions than are required to
+function correctly.
+
+#### Cross-namespace access using a ServiceAccount
```
I will add anchors to all, thanks for the reminder

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,200 @@
+    (not recommended): You can mount service account tokens as Kubernetes
+    Secrets in Pods. These tokens don't expire and don't rotate. This method
+    is not recommended, especially at scale, because of the risks associated
+    with static, long-lived credentials.
```
Done

shannonxtreme commented:
I'm still alive and will work on this today

shannonxtreme commented:
@liggitt @sftim this is ready

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+human users in the cluster. User accounts don't exist in the Kubernetes API
+server; instead, the API server treats user identities as opaque data. You can
+authenticate as a user account using multiple methods.
+
```
Optional note: some Kubernetes distributions / platforms add a custom, extension API that _does_ represent users.

shannonxtreme replied:
Done

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+
+*  **Namespaced:** Each service account is bound to a Kubernetes
+   {{<glossary_tooltip text="namespace" term_id="namespace">}}. Every namespace
+   gets a `default` ServiceAccount upon creation.
```
_optional_
```suggestion
   gets a [`default` ServiceAccount](#default-service-account) upon creation.
```

shannonxtreme replied:
Done

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+* Your Pods need to communicate with the Kubernetes API server, for example in
+  situations such as the following:
+  *  Providing read-only access to sensitive information stored in Secrets.
+  *  Granting [cross-namespace access](TODO), such as allowing a Pod in namespace
```
Highlighting a **TODO**

shannonxtreme replied:
Resolved, thank you

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+  [Authenticating to a private image registry using an `imagePullSecret`](/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account).
+* An external service needs to communicate with the Kubernetes API server. For
+  example, authenticating to the cluster as part of a CI/CD pipeline.
+
```
_optional_
```suggestion
* You use third-party security software in your Kubernetes cluster. The third-party software
  relies on the ServiceAccount identity of different Pods to group those Pods into different
  contexts.

```
[example](https://projectcalico.docs.tigera.io/security/service-accounts)

shannonxtreme replied:
Done (didnt link to example)

sftim replied:
Ah, the example was just for context, not to link to. Thanks. All good here.

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+1. Create the ServiceAccount object in the Kubernetes API using a Kubernetes
+   client like `kubectl` or a manifest that defines the object.
+1. Grant permissions to the ServiceAccount object using an authorization
+   mechanism such as RBAC.
```
_optional_
```suggestion
   mechanism such as [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/).
```

shannonxtreme replied:
Done

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+[projected volume](/docs/concepts/storage/projected-volumes/#serviceaccounttoken).
+In versions earlier than 1.22, Kubernetes provides a long-lived, static token
+to the Pod as a Secret.
+
```
_optional_
```suggestion

<!-- OK to remove this historical detail once Kubernetes 1.31 is released -->

```

because eventually we should assume people are running something vaguely close to being supported.
(IMO, it's fair to impose costs on people who are 6 minor releases out of support).

shannonxtreme replied:
Done and +1

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+## Authenticating service account credentials {#authenticating-credentials}
+
+ServiceAccounts use signed
+{{<glossary_tooltip term_id="jwt" text="JSON Web Tokens (JWTs)">}}
```
I prefer
```suggestion
{{<glossary_tooltip term_id="jwt" text="JSON Web Tokens">}} (JWTs)
```
but not enough to insist on it.

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+
+*   [Learn how to manage your ServiceAccounts as a cluster administrator](/docs/reference/access-authn-authz/service-accounts-admin/).
+*   [Learn how to assign a ServiceAccount to a Pod](/docs/tasks/configure-pod-container/configure-service-account/).
+*   [Read the ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```
Further optional tweak:
```suggestion
* Read the [ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```

shannonxtreme replied:
Ack this

sftim reviewed (changes requested):
Other than a **TODO** that needs fixing, all _my_ feedback here is optional.

You should make sure to address https://github.com/kubernetes/website/pull/38289/files#r1043404706, https://github.com/kubernetes/website/pull/38289/files#r1040600881 and https://github.com/kubernetes/website/pull/38289/files#r1040600078. These got marked as resolved but aren't yet fully addressed.

sftim commented:
/hold cancel

JWT thing is dealt with. I did mark this as _changes requested_ to highlight other feedback that was previously missed.

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+     `example` to read, list, and watch for Lease objects in the
+     `kube-node-lease` namespace.
+* Your Pods need to communicate with an external service. For example,
+  [Authenticating to a private image registry using an `imagePullSecret`](/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account).
```
Done

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+    *   Use service accounts or user accounts created using an external Identity
+        and Access Management (IAM) service, such as from a cloud provider, to
+        authenticate to your cluster.
+
```
Done

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,210 @@
+1. Check the audience claims.
+
+For tokens issued using the `TokenRequest` API, the API server also checks that
+the Pod using the ServiceAccount still exists. For legacy tokens that are
```
Done

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,231 @@
+
+## {{% heading "whatsnext" %}}
+
+*   Learn how to [manage your ServiceAccounts as a cluster administrator](/docs/reference/access-authn-authz/service-accounts-admin/).
+*   Learn how to [assign a ServiceAccount to a Pod](/docs/tasks/configure-pod-container/configure-service-account/).
+*   Read the [ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```
OK, but ideally:
```suggestion
* Learn how to [manage your ServiceAccounts as a cluster administrator](/docs/reference/access-authn-authz/service-accounts-admin/).
* Learn how to [assign a ServiceAccount to a Pod](/docs/tasks/configure-pod-container/configure-service-account/).
* Read the [ServiceAccount API reference](/docs/reference/kubernetes-api/authentication-resources/service-account-v1/).
```

shannonxtreme replied:
Done

shannonxtreme commented:
I see what happened. I Implemented all the previous reviews using Codespaces for the first time and I guess I mucked it up and nothing carried over. I went through every resolved review to make sure that the changes are reflected

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,232 @@
+    * Use service accounts or user accounts created using an external Identity
+      and Access Management (IAM) service, such as from a cloud provider, to
+      authenticate to your cluster.
+    * [Configure the kubelet to retrieve credentials from an image registry](/docs/tasks/administer-cluster/kubelet-credential-provider/).
+    * Use a Device Plugin to access a TPM that allows authentication using
+      a private key.
```
Wording tweak combined with whitespace / list item depth fix.
```suggestion
* [Configure the kubelet to retrieve credentials from an image registry](/docs/tasks/administer-cluster/kubelet-credential-provider/).
* Use a Device Plugin to access a virtual Trusted Platform Module (TPM), that then allows a Pod to authenticate
  using a private key.
```

shannonxtreme replied:
Done

sftim commented:
The whitespace / depth fix in https://github.com/kubernetes/website/pull/38289#discussion_r1072647221 is now essential, sorry, as otherwise we could mislead readers.

shannonxtreme commented:
@sftim  I implemented everything that I could find except for the comment about ServiceAccounts being for more than just Pod processes. I added a suggestion in the comment - do you mind taking a peek when you can?

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,232 @@
+you'd create a RoleBinding object in the `maintenance` namespace to bind the
+Role to the ServiceAccount object. Now, Pods in the `dev` namespace can list
+Job objects in the `maintenance` namespace using that service account.
+
```
:thought_balloon: 

_Doesn't_ block this PR, but you can also allow one ServiceAccount to impersonate another, so there are two options here.

stlaz replied:
Impersonation is always possible, the current documented process appears cleaner to me since you're granting the access explicitly, rather than inheriting ALL the access of the other SA.

shannonxtreme replied:
Let's be prescriptive and document the approach that's "better" from us

sftim commented:
For https://github.com/kubernetes/website/pull/38289#issuecomment-1386007209 are you happy that you know how to move this forward now @shannonxtreme?

shannonxtreme commented:
> For https://github.com/kubernetes/website/pull/38289#issuecomment-1386007209 are you happy that you know how to move this forward now @shannonxtreme?

Aye, I am! I'll commit today ☺️ thank you for the help on Slack!

shannonxtreme commented:
@sftim I implemented the updated intro text!

shannonxtreme commented:
I feel like this page is in a decent spot now

enj commented:
@shannonxtreme does this PR cover #32655 ?

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,237 @@
+
+## What are service accounts? {#what-are-service-accounts}
+
+A service account is a type of non-human account that, in Kubernetes, provides
+a distinct identity in a Kubernetes cluster. Application Pods, system
+components, and entities inside and outside the cluster can use a specific
+ServiceAccount's credentials to identify as that ServiceAccount. This identity
+is useful in various situations, including authenticating to the API server or
+implementing identity-based security policies.
```
I'm happy with this - enough detail not to mislead, concise enough that the reader ought to be comfortable continuing with the rest of the page.

sftim reviewed (commented):
I'd like a tech review from SIG Auth. LGTM for docs.

sftim commented:
> does this PR cover https://github.com/kubernetes/website/issues/32655 ?

Whatever the answer is, that shouldn't block a merge for this PR. We're already tracking the work in #32655 as an issue.

enj commented:
> > does this PR cover #32655 ?
> 
> Whatever the answer is, that shouldn't block a merge for this PR. We're already tracking the work in #32655 as an issue.

Agree, I was more asking if we could close that issue when this PR merged :)

shannonxtreme commented:
It's a good ask @enj and I added a similar line in 143-144 and 165-168. That is the same wording as in the existing doc that Jordan mentioned in #32655. Hope that covers it!

sftim commented:
This might also help with #4410

shannonxtreme commented:
> This might also help with #4410

Would that be a good candidate for use cases in this doc? I think that a brief use case is ok but should link to a new task page or a new section in https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/ for full steps with examples

WDYT @sftim?

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+  *  Granting [cross-namespace access](#cross-namespace), such as allowing a
+     Pod in namespace `example` to read, list, and watch for Lease objects in
+     the `kube-node-lease` namespace.
+* Your Pods need to communicate with an external service. For example, a
+  workload Pod requires an identity for a commercially available cloud API,
+  and the commercial provider allows configuring a suitable trust relationship.
```
This might be a good place to refer to SA token audiences

sftim replied:
OK, but IMO not having it here shouldn't block approval. The perfect is the enemy of the published.

shannonxtreme replied:
I understand the idea but we currently dont have a lot of info in our docs about audiences...I'd like to see better info to link to from here, expanding on https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#serviceaccount-token-volume-projection

WDYT?

stlaz replied:
I think we might often be relying on the readers knowledge of how audiences work, as described in https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation

I added a proposal in https://github.com/kubernetes/website/pull/38289/files#r1099997580 which I think might suffice for this new documentation

shannonxtreme replied:
Done

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+* [Authenticating to a private image registry using an `imagePullSecret`](/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account).
+* An external service needs to communicate with the Kubernetes API server. For
+  example, authenticating to the cluster as part of a CI/CD pipeline.
+* You use third-party security software in your cluster that relies on the
+  ServiceAccount identity of different Pods to group those Pods into different
+  contexts.
```
or perhaps it might be worth mentioning the audiences here?

sftim replied:
I'd be very happy to see a follow-up PR (or more than one) that polishes this page.

Right now we're looking at what's minimum viable to allow a merge.

shannonxtreme replied:
Yes, let's work towards a solid foundation on which we can build and expand. Im more than happy to collab on followups

shannonxtreme replied:
Done because we used @stlaz suggestion for audiences in the 3p section

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+1. Grant permissions to the ServiceAccount object using an authorization
+   mechanism such as
+   [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/).
+1. Assign the ServiceAccount object to Pods during Pod creation.
```
3b. retrieve the ServiceAccount token and use it from an external service

shannonxtreme replied:
DOne

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+ServiceAccount. RBAC lets you define a minimum set of permissions so that the
+service account permissions follow the principle of least privilege. Pods that
+use that service account don't get more permissions than are required to
+function correctly.
```
This might also be a good place to add the reference from earlier: `/docs/reference/access-authn-authz/rbac/#service-account-permissions`

shannonxtreme replied:
Done

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+To assign a ServiceAccount to a Pod, you set the `spec.serviceAccountName`
+field in the Pod specification. Kubernetes then automatically provides the
+credentials for that ServiceAccount to the Pod. In v1.22 and later, Kubernetes
+gets a short-lived, automatically rotating token using the `TokenRequest` API
```
It might make sense to highlight the `automatically rotating` here to outline that the workload must consider changes in the token file.

shannonxtreme replied:
Done

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+location, or for an audience that isn't the API server, use one of the
+following methods:
+
+* [TokenRequest API](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
+  (recommended): Request a short-lived service account token using your
+  *application code*. The token expires automatically and can rotate upon
+  expiration.
+* [Token Volume Projection](/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection)
+  (recommended): In Kubernetes v1.20 and later, use the Pod specification to
+  tell the kubelet to add the service account token to the Pod as a
+  *projected volume*. Projected tokens expire automatically, and the kubelet
+  rotates the token before it expires. The ServiceAccount token is also
+  invalidated in the Kubernetes API when the Pod or the ServiceAccount is
+  deleted.
+* [Service Account Token Secrets](/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token)
+  (not recommended): You can mount service account tokens as Kubernetes
+  Secrets in Pods. These tokens don't expire and don't rotate. This method
+  is not recommended, especially at scale, because of the risks associated
+  with static, long-lived credentials. In Kubernetes v1.24 and later, the
+  [LegacyServiceAccountTokenNoAutoGeneration feature gate](/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-graduated-or-deprecated-features)
+  prevents Kubernetes from automatically creating these tokens for
+  ServiceAccounts.
```
should we mention `kubectl create token` here, or update any of the linked docs?

sftim replied:
IMO: that can happen in a follow up PR.

stlaz replied:
That's fair.

I should just add that if we wanted to keep it in this doc, the right place would be around the TokenRequest API as that's used as the backend for the command.

shannonxtreme replied:
Ack, thank you

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+ServiceAccounts use signed
+{{<glossary_tooltip term_id="jwt" text="JSON Web Tokens">}}  (JWTs)
+to authenticate to the Kubernetes API server, and to any other system where a
+trust relationship exists. Depending on the token issuer, a ServiceAccount
```
Isn't the kube-apiserver the only available issuer today? Is there a scenario where any of the following claims would be missing?

liggitt replied:
Yeah, this depends less on the issuer and more on the type of token. Secret-based non-expiring tokens don't have an expiry or audience claim, new time-bound tokens do.

shannonxtreme replied:
Is there something to document here about that? Seems like a legacy behavior at this point

stlaz replied:
We may want to distinguish based on the type of the token rather than the issuer, something like
```suggestion
trust relationship exists. Depending on how the token was issued (legacy, secret-based token or the newer time-based), a ServiceAccount
```

sftim replied:
Tip: list the current mechanism first:
```suggestion
trust relationship exists. Depending on how the token was issued (either time-limited using
a [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/),
or the legacy mechanism using a Secret), a ServiceAccount
```

liggitt replied:
@sftim's edit lgtm

shannonxtreme replied:
Done

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+
+1. Check the token signature.
+1. Check whether the token has expired.
+1. Check whether the token is currently valid.
```
This is rather vague, all the other points are a also a check that the token is currently valid. Did you mean any specific check?

Tokens can  be (and very often are) bound to a certain API object. We may want to mention that check here, too.

sftim replied:
What are the checks that are actually appropriate @stlaz? That information will help @shannonxtreme tweak the doc.

stlaz replied:
I see only Pod and Secret are currenly allowed to be referenced.

The code for this specific authentication lives in https://github.com/kubernetes/kubernetes/blob/8a72f059997c548c5432980682f50ac7bfa036ae/pkg/serviceaccount/jwt.go#L262

along with the validator in https://github.com/kubernetes/kubernetes/blob/8a72f059997c548c5432980682f50ac7bfa036ae/pkg/serviceaccount/claims.go#L111

liggitt replied:
I would hesitate to enumerate all the "token is currently valid" checks the API server does here in a way that encourages other parties to replicate those checks outside the API server, but something like "Checks whether object references in the token claims are still valid" could work

shannonxtreme replied:
Added this suggestion

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+1. Check whether the token is currently valid.
+1. Check the audience claims.
+
+For tokens issued using the `TokenRequest` API, the API server also checks that
+the specific Pod using the ServiceAccount still exists, matching by the
```
I see, the bound object is mentioned here. The bound object API of the token is not limited to pods only - https://kubernetes.io//docs/reference/kubernetes-api/authentication-resources/token-request-v1/#TokenRequestSpec

shannonxtreme replied:
I added Jordan's suggestion "Checks whether object references in the token claims are still valid". And changed Pod in the following paragraph to "referenced object"

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+For more information about the authentication process, refer to
+[Authentication](/docs/reference/access-authn-authz/authentication/#service-account-tokens).
+
+### Authenticating service account credentials in your own code {authenticating-in-code}
```
This section definitely should mention token audiences.

shannonxtreme replied:
Could you suggest a line or two? We want them to specify intended audiences for tokens obtained using TokenRequest so that the token authenticator can validate that the recipient is in the list of audiences/

sftim replied:
Wrapped, that becomes
```markdown
Your application should always define its own audience that it is going to accept, and it should
check that the token's audiences match the audiences the application expects. This is to help
minimize the scope of the token, making it only usable in your application and nowhere else.
```

shannonxtreme replied:
Done

stlaz commented on content/en/docs/reference/glossary/jwt.md:
```diff
@@ -0,0 +1,21 @@
+date: 2023-01-17
+full_link: https://www.rfc-editor.org/rfc/rfc7519
+short_description: >
+  An authentication token format based on JSON data. JWTs can be digitally
+  signed.
```
They can also be encrypted. But they are not an _authentication_ token format, they are "means of representing claims to be transferred between two parties", to quote the linked RFC.

I would remove the second sentence and just replace the original description from the one from the RFC.

shannonxtreme replied:
Done, although I expanded the "more" section to explain how Kubernetes typically uses it

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+* Issue your own tokens using another mechanism, and then use
+  [Webhook Token Authentication](/docs/reference/access-authn-authz/authentication/#webhook-token-authentication)
+  to validate bearer tokens using your own validation service.
+* Provide your own identities to Pods.
+    * [Use the SPIFFE CSI driver plugin to provide SPIFFE SVIDs as X.509 certificate pairs to Pods](https://cert-manager.io/docs/projects/csi-driver-spiffe/).
+    * [Use a service mesh such as Istio to provide certificates to Pods](https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/).
```
is it ok to mention 3rd parties here?

sftim replied:
Yes, it is. We should use a thirty party content shortcode to mark that we're doing so.

shannonxtreme replied:
Done

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+      authenticate to your cluster.
+* [Configure the kubelet to retrieve credentials from an image registry](/docs/tasks/administer-cluster/kubelet-credential-provider/).
+* Use a Device Plugin to access a virtual Trusted Platform Module (TPM), which
+  then allows authentication using a private key.
```
+ CSR API with client certs, possibly?

sftim replied:
OK, but IMO not having it here shouldn't block approval. The perfect is the enemy of the published.

shannonxtreme replied:
Added with a link to https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/

stlaz commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+      and Access Management (IAM) service, such as from a cloud provider, to
+      authenticate to your cluster.
+* [Configure the kubelet to retrieve credentials from an image registry](/docs/tasks/administer-cluster/kubelet-credential-provider/).
+* Use a Device Plugin to access a virtual Trusted Platform Module (TPM), which
```
no references here?

sftim replied:
That's OK.

stlaz replied:
Could we please at least link https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/ ? I was not aware of it before.

stlaz reviewed (commented):
This is a really good summary for ServiceAccounts :rocket:  I added a few comments of how I think we might improve it even further.

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+| Location | Kubernetes API (ServiceAccount object) | External |
+| Scope | Namespace | Cluster |
+| Access control | Kubernetes RBAC or other [authorization mechanisms](/docs/reference/access-authn-authz/authorization/#authorization-modules) | Kubernetes RBAC or other identity and access management mechanisms |
+| Intended use | Pods | People |
```
“Services” might make readers thing about the kind of Service that manages EndpointSlices, so let's avoid that too.

Maybe “apps” or “code”?

stlaz replied:
Yes, I was wondering whether there would be a better way to describe it. I think "Apps" or "Applications"  would be the most fitting.

shannonxtreme replied:
Workloads?

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+When you create a cluster, Kubernetes automatically creates a ServiceAccount
+object named `default` for every namespace in your cluster. The `default`
+service accounts in each namespace get no permissions by default other than the
+[default API discovery permissions](/docs/reference/access-authn-authz/rbac/#service-account-permissions)
```
We should try to be slightly neutral about how `system:authenticated` gets permissions (it could be RBAC, it could be another mechanism).

shannonxtreme replied:
Since RBAC is the "built-in" mechanism, I don't have an issue with changing the link to the suggestion. @sftim let me know which you think is better for here

sftim replied:
RBAC is in-tree but is not part of what we consider as _certified Kubernetes_.

shannonxtreme replied:
```suggestion
[default API discovery permissions](/docs/reference/access-authn-authz/rbac/#service-account-permissions)
```

that Kubernetes grants to all authenticated accounts using an access control
mechanism such as role-based access control (RBAC).

shannonxtreme replied:
Implemented Tim's with minor tweaks

liggitt commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+If you deploy a Pod in a namespace, and you don't manually assign a
+ServiceAccount object to the Pod, Kubernetes assigns the `default` service
+account for that namespace to the Pod and provides the credentials for that
+ServiceAccount to the Pod. To prevent Kubernetes from automatically providing
+credentials for the `default` ServiceAccount, set the
```
```suggestion
ServiceAccount to the Pod. To prevent Kubernetes from automatically injecting
credentials for the configured ServiceAccount into the Pod, set the
```

shannonxtreme replied:
I reworded this paragraph to make more sense. It originally focused only on the default account, so talking suddenly about configured accounts would have been confusing. PTAL and resolve if ok with you

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+| Location | Kubernetes API (ServiceAccount object) | External |
+| Scope | Namespace | Cluster |
+| Access control | Kubernetes RBAC or other [authorization mechanisms](/docs/reference/access-authn-authz/authorization/#authorization-modules) | Kubernetes RBAC or other identity and access management mechanisms |
+| Intended use | Pods | People |
```
I'd go for:
```markdown
| Intended use | Pods, automation | people |
```

so that the initial caps helps to mark out API kinds

shannonxtreme replied:
In this table I dont think the caps differentiation is needed for People, it's a different column. Lets go with "Workloads, automation" and "People"

I don't know if "Automation" without expanding on what is being automated makes things less clear though

sftim commented:
Approval stands, but
/hold

to highlight that https://github.com/kubernetes/website/pull/38289#pullrequestreview-1287149259 is important.

@shannonxtreme you can unhold this once you've accounted for that and have got LGTM.

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+   {{<glossary_tooltip text="namespace" term_id="namespace">}}. Every namespace
+   gets a [`default` ServiceAccount](#default-service-account) upon creation.
+
+*  **Lightweight:** Service accounts exist in the cluster and use the structure
```
It's supposed to mean that it's defined in the Kubernetes API. I reworded to "and are defined in the Kubernetes API"

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+| Description | ServiceAccount | User or group |
+| --- | --- | --- |
+| Location | Kubernetes API (ServiceAccount object) | External |
+| Scope | Namespace | Cluster |
```
Let's remove this row, I don't think that it works unless I change Scope to something that means "Location of original object" which is not a valuable bit of info at this stage

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+| Description | ServiceAccount | User or group |
+| --- | --- | --- |
+| Location | Kubernetes API (ServiceAccount object) | External |
+| Scope | Namespace | Cluster |
```
Let's remove this row, I don't think that it works unless I change Scope to something that means "Location of original object" which is not a valuable bit of info at this stage

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+If you deploy a Pod in a namespace, and you don't manually assign a
+ServiceAccount object to the Pod, Kubernetes assigns the `default` service
+account for that namespace to the Pod and provides the credentials for that
+ServiceAccount to the Pod. To prevent Kubernetes from automatically providing
+credentials for the `default` ServiceAccount, set the
+`automountServiceAccountToken` field in your Pod specification to `false`.
```
I reworded this paragraph to make more sense. It originally focused only on the default account, so talking suddenly about configured accounts would have been confusing. PTAL and resolve if ok with you

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,241 @@
+{{<glossary_tooltip term_id="jwt" text="JSON Web Tokens">}}  (JWTs)
+to authenticate to the Kubernetes API server, and to any other system where a
+trust relationship exists. Depending on the token issuer, a ServiceAccount
+token might also have an expiry time, an audience, and a time after which the
```
@stlaz we do mention audience here FYI

shannonxtreme commented:
@stlaz and @liggitt I've made changes based on your reviews and mentioned in comments whether you need to take another look at a section.  Please do and either respond or LGTM :)

shannonxtreme commented:
Preview: https://deploy-preview-38289--kubernetes-io-main-staging.netlify.app/docs/concepts/security/service-accounts/

sftim commented:
/hold cancel

stlaz reviewed (commented):
great work, only a few more minor comments

sftim commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,249 @@
+object named `default` for every namespace in your cluster. The `default`
+service accounts in each namespace get no permissions by default other than the
+[default API discovery permissions](/docs/reference/access-authn-authz/rbac/#service-account-permissions)
+that Kubernetes grants to all authenticated accounts. If you delete the
```
```suggestion
that Kubernetes grants to all authenticated principals whenever Kubernetes RBAC is
enabled. If you delete the
```
maybe?

shannonxtreme replied:
Went with "to all authenticated principals if RBAC is enabled.

shannonxtreme replied:
Done

shannonxtreme commented:
Reviews be like

![breakfast-secondbreakfast.gif](https://user-images.githubusercontent.com/30392854/218471051-584b54d4-9645-4186-9d52-b553d504c11f.gif)


Thank you for the prompt edits y'all, I'll implement today!

sftim commented:
Looking forward to seeing this one merge.

liggitt reviewed (commented):
all the updates lgtm, https://github.com/kubernetes/website/pull/38289/files#r1104411643 is the only open thread I had commented on and @sftim's suggested edit looked good

shannonxtreme commented on content/en/docs/concepts/security/service-accounts.md:
```diff
@@ -0,0 +1,249 @@
+with the credentials for an assigned ServiceAccount, whether that is the
+`default` ServiceAccount or a custom ServiceAccount that you specify. 
+
+To prevent Kubernetes from automatically injecting
+credentials for a specified ServiceAccount or the `default` ServiceAccount, set the
+`automountServiceAccountToken` field in your Pod specification to `false`.
```
done

shannonxtreme commented:
I think I've resolved everything that needed resolving @stlaz and @liggitt

liggitt commented:
/lgtm
